fawltydeps/extract_declared_dependencies.py

```diff
@@ -288,27 +288,32 @@ def parse_pyproject_toml(path: Path) -> Iterator[DeclaredDependency]:
     else:
         logger.debug("%s does not contain [tool.poetry].", source)
 
-    if "dynamic" in parsed_contents.get("tool", {}).get("setuptools", {}).keys():
-        dynamic_section = parsed_contents["tool"]["setuptools"]["dynamic"]
-        dynamic_deps_files = dynamic_section.get("dependencies", {}).get("file", [])
-        dynamic_optional_deps_files = [
-            file_list
-            for optional_deps_item in [
-                optional_deps_section.get("file", {})
-                for optional_deps_section in dynamic_section.get(
-                    "optional-dependencies", {}
-                ).values()
+    if "dynamic" in parsed_contents.get("project", {}) and (
+        "dependencies" in parsed_contents["project"]["dynamic"]
+        or "optional-dependencies" in parsed_contents["project"]["dynamic"]
+    ):
+        if "dynamic" in parsed_contents.get("tool", {}).get("setuptools", {}):
+            dynamic_section = parsed_contents["tool"]["setuptools"]["dynamic"]
+            dynamic_deps_files = dynamic_section.get("dependencies", {}).get("file", [])
```

Comment:
I feel this is getting hard to read. Maybe adding a comment before each of these conditions to indicate what we're trying to do would help?

Reply:
This is one of those cases where I feel that looking things up directly and handling the resulting exceptions (aka. "ask forgiveness rather than permission") leads to nicer code (less repetition and nesting of `if`s):

```suggestion
    try:
        dynamic = parsed_contents["project"]["dynamic"]
    except KeyError:
        dynamic = []

    deps_files = []
    try:
        if "dependencies" in dynamic:
             deps_files = parsed_contents["tool"]["setuptools"]["dynamic"]["dependencies"]["file"]
    except KeyError:
        pass

    optional_deps_files = {}
    try:
        if "optional-dependencies" in dynamic:
             optional_deps = parsed_contents["tool"]["setuptools"]["dynamic"]["optional-dependencies"]
             optional_deps_files = {k: v["file"] for k, v in optional_deps.items()}
    except KeyError:
        pass

    dynamic_files = deps_files + optional_deps_files
    for req_file in dynamic_files:
        ...
```

Also, looking at the rest of this function, I wonder if it makes sense to put all the new/dynamic code into a seprarate helper, named `parse_dynamic_pyproject_contents()` or similar.

(Finally I also suspect we'd need to select between calling this helper, and calling `parse_pep621_pyproject_contents` depending on the contents of the `project.dynamic` section...)

Reply:
Addressed in https://github.com/tweag/FawltyDeps/pull/351/commits/4242069da6847018533f95b1a4af1ee86159a4a0 and https://github.com/tweag/FawltyDeps/pull/351/commits/4c350f5438802703a55cbe67bc02ca452aef021d